File: experiments/runs/2026-09-14_14-28-42_claim-office-example-mapping_external-tcrdd-bsene-2026-09-14-pi_gpt-5-6-sol-codex/src/office.ts
export interface Scenario {
  customer: { yearsWithMHPCO: number };
  steps: Array<{ op: 'quote'; items: Item[] } | ClaimStep>;
}

export interface Item {
  type: string;
  material?: string;
  enchantment?: number;
  cursed?: boolean;
}

export interface ClaimStep {
  op: 'claim';
  policy: number;
  incident: { cause: string; damages: Array<{ itemType: string; amount: number }> };
}

export type Result = { premium: number } | { payout: number; remainingCap: number };

type Policy = { items: Item[]; remainingCap: number };

const COMPONENT_BLOCK_SIZE = 3;
const COMPONENT_BLOCK_SAVING = 15;
const CURSE_DIVISOR = 2;
const ENCHANTMENT_PREMIUM_THRESHOLD = 5;
const HIGH_ENCHANTMENT_CLAIM_THRESHOLD = 8;
const PERCENT_DENOMINATOR = 100;
const HIGH_ENCHANTMENT_PERCENT = 30;
const LOYALTY_PERCENT = 20;
const INITIAL_ASSESSMENT_PERCENT = 10;
const FOLLOW_UP_PERCENT = 15;
const PROCESSING_FEE = 5;
const DEDUCTIBLE = 100;
const CAP_MULTIPLIER = 2;
const LOYALTY_YEARS = 2;

const VALUES: Record<string, number> = {
  sword: 1000,
  amulet: 600,
  staff: 800,
  potion: 400,
  rune: 250,
  moonstone: 250,
};

const PREMIUMS: Record<string, number> = {
  sword: 100,
  amulet: 60,
  staff: 80,
  potion: 40,
  rune: 25,
  moonstone: 25,
};

function quotePremium(items: Item[], loyal: boolean, followUp: boolean): number {
  const componentCounts = items.reduce<Record<string, number>>((counts, item) => {
    counts[item.type] = (counts[item.type] ?? 0) + 1;
    return counts;
  }, {});
  const base = items.reduce((total, item) => total + PREMIUMS[item.type], 0)
    - (componentCounts.rune === COMPONENT_BLOCK_SIZE ? COMPONENT_BLOCK_SAVING : 0)
    - (componentCounts.moonstone === COMPONENT_BLOCK_SIZE ? COMPONENT_BLOCK_SAVING : 0);
  const riskSurcharges = items.reduce((total, item) => {
    const itemBase = PREMIUMS[item.type];
    return total + (item.cursed ? itemBase / CURSE_DIVISOR : 0)
      + ((item.enchantment ?? 0) >= ENCHANTMENT_PREMIUM_THRESHOLD
        ? itemBase * HIGH_ENCHANTMENT_PERCENT / PERCENT_DENOMINATOR : 0);
  }, 0);
  const loyaltyDiscount = loyal ? base * LOYALTY_PERCENT / PERCENT_DENOMINATOR : 0;
  const followUpDiscount = followUp ? base * FOLLOW_UP_PERCENT / PERCENT_DENOMINATOR : 0;
  const initialAssessment = base * INITIAL_ASSESSMENT_PERCENT / PERCENT_DENOMINATOR;
  return Math.ceil(base + riskSurcharges + initialAssessment
    - loyaltyDiscount - followUpDiscount + PROCESSING_FEE);
}

function processClaim(policy: Policy, step: ClaimStep): Result {
  const remainingItems = [...policy.items];
  const desired = step.incident.damages.reduce((sum, damage) => {
    if (damage.amount < 0) throw new Error('Negative damage amount');
    const itemIndex = remainingItems.findIndex(item => item.type === damage.itemType);
    if (itemIndex < 0) throw new Error(`Item is not covered: ${damage.itemType}`);
    const item = remainingItems.splice(itemIndex, 1)[0];
    const reimbursable = (item.enchantment ?? 0) >= HIGH_ENCHANTMENT_CLAIM_THRESHOLD
      ? damage.amount / CURSE_DIVISOR : damage.amount;
    return sum + Math.max(0, reimbursable - DEDUCTIBLE);
  }, 0);
  const payout = Math.floor(Math.min(desired, policy.remainingCap));
  policy.remainingCap -= payout;
  return { payout, remainingCap: policy.remainingCap };
}

export function processScenario(scenario: Scenario): { results: Result[] } {
  let quotes = 0;
  const policies = new Map<number, Policy>();
  const results = scenario.steps.map<Result>((step, stepIndex) => {
    if (step.op === 'quote') {
      for (const item of step.items) {
        if (!(item.type in PREMIUMS)) throw new Error(`Unknown item type: ${item.type}`);
      }
      const result = {
        premium: quotePremium(step.items, scenario.customer.yearsWithMHPCO >= LOYALTY_YEARS, quotes > 0),
      };
      quotes += 1;
      const insuranceSum = step.items.reduce((sum, item) => sum + VALUES[item.type], 0);
      policies.set(stepIndex, { items: step.items, remainingCap: insuranceSum * CAP_MULTIPLIER });
      return result;
    }
    const policy = policies.get(step.policy);
    if (!policy) throw new Error('Claim requires a policy');
    return processClaim(policy, step);
  });
  return { results };
}

File: experiments/runs/2026-09-21_22-14-14_claim-office-python-example-mapping_exact-ptdd-v1-pi_gpt-5-6-sol-codex/src/office.py
import math


PROCESSING_FEE = 5
DEDUCTIBLE = 100
CAP_MULTIPLIER = 2
CLAIM_ENCHANTMENT_THRESHOLD = 8
HIGH_ENCHANTMENT_REIMBURSEMENT_RATE = 0.50
BASE_PREMIUM = {"sword": 100, "amulet": 60, "staff": 80, "potion": 40, "rune": 25, "moonstone": 25}
INSURANCE_VALUE = {
    "sword": 1000,
    "amulet": 600,
    "staff": 800,
    "potion": 400,
    "rune": 250,
    "moonstone": 250,
}
INITIAL_ASSESSMENT_RATE = 0.10
CURSE_SURCHARGE_RATE = 0.50
ENCHANTMENT_SURCHARGE_RATE = 0.30
PREMIUM_ENCHANTMENT_THRESHOLD = 5
LOYALTY_YEARS_THRESHOLD = 2
LOYALTY_DISCOUNT_RATE = 0.20
FOLLOW_UP_DISCOUNT_RATE = 0.15
COMPONENT_BLOCK_SIZE = 3
COMPONENT_BLOCK_SAVING = 15
COMPONENT_TYPES = ("rune", "moonstone")


def policy_base_premium(items):
    base = sum(BASE_PREMIUM[item["type"]] for item in items)
    for component_type in COMPONENT_TYPES:
        if sum(item["type"] == component_type for item in items) == COMPONENT_BLOCK_SIZE:
            base -= COMPONENT_BLOCK_SAVING
    return base


def item_risk_surcharge(item):
    rate = CURSE_SURCHARGE_RATE if item.get("cursed", False) else 0
    if item.get("enchantment", 0) >= PREMIUM_ENCHANTMENT_THRESHOLD:
        rate += ENCHANTMENT_SURCHARGE_RATE
    return BASE_PREMIUM[item["type"]] * rate


def policy_wide_adjustment(base, years_with_mhpco, is_follow_up):
    adjustment = base * INITIAL_ASSESSMENT_RATE
    if years_with_mhpco >= LOYALTY_YEARS_THRESHOLD:
        adjustment -= base * LOYALTY_DISCOUNT_RATE
    if is_follow_up:
        adjustment -= base * FOLLOW_UP_DISCOUNT_RATE
    return adjustment


def quote_premium(items, years_with_mhpco=0, is_follow_up=False):
    base = policy_base_premium(items)
    risk_surcharge = sum(item_risk_surcharge(item) for item in items)
    adjustment = policy_wide_adjustment(base, years_with_mhpco, is_follow_up)
    return math.ceil(base + risk_surcharge + adjustment + PROCESSING_FEE)


def create_policy(items):
    insurance_sum = sum(INSURANCE_VALUE[item["type"]] for item in items)
    return {"items": items, "remaining_cap": insurance_sum * CAP_MULTIPLIER}


def damage_reimbursement(item, damage):
    amount = damage["amount"]
    if item.get("enchantment", 0) >= CLAIM_ENCHANTMENT_THRESHOLD:
        amount *= HIGH_ENCHANTMENT_REIMBURSEMENT_RATE
    return max(amount - DEDUCTIBLE, 0)


def match_damage_items(policy, damages):
    available_items = list(policy["items"])
    matches = []
    for damage in damages:
        item = next(
            (
                item
                for item in available_items
                if item["type"] == damage["itemType"]
            ),
            None,
        )
        if item is None:
            raise ValueError("damage item is not covered by the policy")
        available_items.remove(item)
        matches.append((item, damage))
    return matches


def validate_damage_amounts(damages):
    if any(damage["amount"] < 0 for damage in damages):
        raise ValueError("damage amount cannot be negative")


def claim_policy(policy, damages):
    validate_damage_amounts(damages)
    matches = match_damage_items(policy, damages)
    desired = sum(damage_reimbursement(item, damage) for item, damage in matches)
    payout = math.floor(min(desired, policy["remaining_cap"]))
    policy["remaining_cap"] -= payout
    return {"payout": payout, "remainingCap": policy["remaining_cap"]}
